File: Problem_2.java
// Time Complexity : O(n*n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// take two pointer and compare from each char index, for even more the one pointer towards rigt (i+1)
// Your code here along with comments explaining your approach
class Solution {
    int start; int end;
    public String longestPalindrome(String s) {
        if( s== null || s.length() < 2) return s;
        int n = s.length();
        for(int i = 0; i < n; i++){
            pointer(s,i,i); // odd case
            pointer(s,i,i+1); // odd case
        }
        return s.substring(start, end+1);
    }
    private void pointer(String s, int l, int r){
        while( l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)){
            l--; r++;
        }
        // left is more lefter and right is more righter
        l++; r--;
        if( r - l > end - start){
            start = l;
            end = r;
        }
    }
}

// DP - solution
// Time Complexity : O(n*n)
// Space Complexity :O(n*n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// take a string to string matrix and check if both character is same, if yes then check if they are only two character to compare ( i - j > 2) , or there inncer string is palindrome or not by suberray dp[i-1][j+1].
// Your code here along with comments explaining your approach
// class Solution {
//     public String longestPalindrome(String s) {
//         if( s == null || s.length() < 2 ) return s;
//         int n = s.length();
//         boolean [][] dp = new boolean [n][n];
//         int start = 0;
//         int end = 0;
//         for(int i = 0; i < n; i++){
//             for(int j = 0; j <= i; j++){
//                 if(s.charAt(i) == s.charAt(j) && ( (i - j <2) || dp[i-1][j+1])){
//                      dp[i][j] = true;
//                      if( i - j > end - start){
//                          start = j;
//                          end = i;
//                      }
//                 }
//             }
//         }
//         return s.substring(start,end+1);
//     }
// }

File: Problem_2.js
// Problem2: (https://leetcode.com/problems/longest-palindromic-substring/)

// TC: O(n^2)
// SC: O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


let start, end, max;

var expandAroundCenter = (s, left, right) => {
    let n = s.length;
    while (left >= 0 && right < n && s[left] === s[right]) {
        left--;
        right++;
    }
    left++;
    right--;
    if (right - left + 1 > max) {
        max = right - left + 1;
        start = left;
        end = right;
    }
}
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
    if (s === null || s.length < 2)
        return s;

    start = 0;
    end = 0;
    max = 0;

    for (let i = 0; i < s.length; i++) {
        expandAroundCenter(s, i, i);
        if (i + 1 < s.length)
            expandAroundCenter(s, i, i + 1);
    }
    return s.slice(start, end + 1);
};